lib/nerves_network/dhcp_manager.ex

```diff
@@ -55,216 +37,223 @@ defmodule Nerves.Network.DHCPManager do
-
-  @spec consume(context, event, t) :: t
   defp consume(_, :noop, state), do: state
-
   ## Context: :removed
   defp consume(:removed, :ifadded, state) do
-    :ok = Nerves.NetworkInterface.ifup(state.ifname)
+    case Nerves.NetworkInterface.ifup(state.ifname) do
+      :ok ->
+        {:ok, status} = Nerves.NetworkInterface.status state.ifname
+        notify(Nerves.NetworkInterface, state.ifname, :ifchanged, status)
 
-    {:ok, status} = Nerves.NetworkInterface.status state.ifname
-    notify(Nerves.NetworkInterface, state.ifname, :ifchanged, status)
+        state
+          |> goto_context(:down)
+      {:error, _} ->
+        # The interface isn't quite up yet. Retry
+        Process.send_after self(), :retry_ifadded, 250
+        state
+          |> goto_context(:retry_add)
+    end
+  end
+  defp consume(:removed, :retry_ifadded, state), do: state
+  defp consume(:removed, :ifdown, state), do: state
 
-    state |> goto_context(:down)
+  ## Context: :retry_add
+  defp consume(:retry_add, :ifremoved, state) do
+    state
+      |> goto_context(:removed)
   end
+  defp consume(:retry_add, :retry_ifadded, state) do
+    {:ok, status} = Nerves.NetworkInterface.status(state.ifname)
+    notify(Nerves.NetworkInterface, state.ifname, :ifchanged, status)
 
-  defp consume(:removed, :ifdown, state), do: state
+    state
+      |> goto_context(:down)
+  end
 
   ## Context: :down
   defp consume(:down, :ifadded, state), do: state
-
   defp consume(:down, :ifup, state) do
     state
-      |> start_udhcpc
+      |> start_dhclient
       |> goto_context(:dhcp)
   end
-
   defp consume(:down, :ifdown, state) do
     state
-      |> stop_udhcpc
+      |> stop_dhclient
   end
-
   defp consume(:down, :ifremoved, state) do
     state
-      |> stop_udhcpc
+      |> stop_dhclient
       |> goto_context(:removed)
   end
 
   ## Context: :dhcp
   defp consume(:dhcp, :ifup, state), do: state
-
   defp consume(:dhcp, {:deconfig, _info}, state), do: state
 
   defp consume(:dhcp, {:bound, info}, state) do
+    Logger.debug fn -> "#{__MODULE__}: consume :bound info: #{inspect info}" end
     state
       |> configure(info)
       |> goto_context(:up)
   end
 
+
   defp consume(:dhcp, {:leasefail, _info}, state) do
     dhcp_retry_timer = Process.send_after(self(), :dhcp_retry, state.dhcp_retry_interval)
     %{state | dhcp_retry_timer: dhcp_retry_timer}
-      |> stop_udhcpc
+      |> stop_dhclient
       |> start_link_local
       |> goto_context(:up)
-  end
 
+  end
   defp consume(:dhcp, :ifdown, state) do
     state
-      |> stop_udhcpc
+      |> stop_dhclient
       |> goto_context(:down)
   end
 
   ## Context: :up
   defp consume(:up, :ifup, state), do: state
-
   defp consume(:up, :dhcp_retry, state) do
     state
-      |> start_udhcpc
+      |> start_dhclient
       |> goto_context(:dhcp)
   end
-
   defp consume(:up, :ifdown, state) do
     state
-      |> stop_udhcpc
+      |> stop_dhclient
```

Comment:
Same here... We might want some fancy way of switching between udhcpc and dhclient at the compile time... It is a nice-to-have. Perhaps the nerves.network folks could pick this burden...

Reply:
What is the reason for a strong 'changes reqeusted?'